### PR TITLE
Advanced AVM features

### PR DESCRIPTION
**Pull request by robdmoore**
## Proposed Changes

- Added the ability to specify a rekey for app call transactions
- Updated `assetBulkOptIn` and `assetBulkOptOut` to conform to algokit utils standards (**note:** includes breaking changes)
- Added `assetOptIn` and `assetOptOut` functions
- Added algokit.rekeyAccount function
- Added lease support to all transfer functions and added convenient lease encoding
- Docs updates

### CODE REVIEW

File: src/testing/_asset.ts [hunk 1]
```diff
@@ -1,6 +1,4 @@
-import { Account, Algodv2, Kmd, makeAssetCreateTxnWithSuggestedParamsFromObject } from 'algosdk'
-import { algos, microAlgos } from '../amount'
-import { ensureFunded } from '../transfer'
+import { Account, Algodv2, makeAssetCreateTxnWithSuggestedParamsFromObject } from 'algosdk'
```
Comment (robdmoore) [inline src/testing/_asset.ts:1]: Renamed this file to `_asset.ts` to reflect the fact that we don't want to export these methods outside of algokit

File: src/app.ts [hunk 2]
```diff
@@ -614,12 +615,12 @@ export async function getAppArgsForABICall(args: ABIAppCallArgs, from: SendTrans
     sender: getSenderAddress(from),
     signer: signer,
     boxes: args.boxes?.map(getBoxReference),
-    lease: typeof args.lease === 'string' ? encoder.encode(args.lease) : args.lease,
+    lease: encodeLease(args.lease),
```
Comment (neilcampbell) [inline src/app.ts:618]: `const encoder = new TextEncoder()` on line 591 can be removed now

File: src/asset.spec.ts [hunk 3]
```diff
@@ -1,121 +1,103 @@
...
+    const secondAccount = await generateAccount({ initialFunds: (1).algos() })
 
-    await ensureFunds(algod, secondAccount, kmd)
-    await optIn(algod, secondAccount, dummyAssetIds)
+    await algokit.assetBulkOptIn({ account: secondAccount, assetIds: dummyAssetIds }, algod)
 
     const secondAccountInfo = await algod.accountInformation(secondAccount.addr).do()
     expect(secondAccountInfo['total-assets-opted-in']).toBe(2)
 
     // await optIn(algod, secondAccount, dummyAssetIds)
```
Comment (neilcampbell) [inline src/asset.spec.ts:37]:
```suggestion
```

File: src/transaction.ts [hunk 4]
```diff
@@ -52,6 +54,42 @@ export function encodeTransactionNote(note?: TransactionNote): Uint8Array | unde
...
+/** Encodes a transaction lease into a 32-byte array ready to be included in an Algorand transaction.
+ *
+ * @param lease The transaction lease as a string or binary array or null/undefined if there is no lease
+ * @returns the transaction lease ready for inclusion in a transaction or `undefined` if there is no lease
+ * @throws if the length of the data is > 32 bytes or empty
+ * @example algokit.encodeLease('UNIQUE_ID')
+ * @example algokit.encodeLease(new Uint8Array([1, 2, 3]))
+ */
+export function encodeLease(lease?: string | Uint8Array): Uint8Array | undefined {
+  if (lease == null || typeof lease === 'undefined') {
```
Comment (neilcampbell) [inline src/transaction.ts:66]:
```suggestion
  if (lease === null || typeof lease === 'undefined') {
```

**Review by neilcampbell — APPROVED** [review]
Looks good!

**Review by robdmoore — COMMENTED** [review]
Review suggestions

**Pull request merged**